stix2/registration.py

```diff
@@ -146,30 +159,25 @@ def _register_extension(
     """
     ext_type = new_extension._type
 
-    # Need to check both toplevel and nested properties
-    prop_groups = [new_extension._properties]
-    if hasattr(new_extension, "_toplevel_properties"):
-        prop_groups.append(new_extension._toplevel_properties)
-    prop_names = itertools.chain.from_iterable(prop_groups)
-
     _validate_type(ext_type, version)
-
-    if not new_extension._properties:
-        raise ValueError(
-            "Invalid extension: must define at least one property: " +
-            ext_type,
-        )
-
     if version == "2.1":
         if not (ext_type.endswith('-ext') or ext_type.startswith('extension-definition--')):
             raise ValueError(
                 "Invalid extension type name '%s': must end with '-ext' or start with 'extension-definition--<UUID>'." %
                 ext_type,
             )
 
-        for prop_name in prop_names:
-            if not re.match(PREFIX_21_REGEX, prop_name):
-                raise ValueError("Property name '%s' must begin with an alpha character." % prop_name)
+    # Need to check both toplevel and nested properties
+    combined_props = {
+        **new_extension._properties,
+        **getattr(new_extension, "_toplevel_properties", dict()),
+    }
+    if not combined_props:
+        raise ValueError(
+            "Invalid extension: must define at least one property: " +
+            ext_type,
+        )
+    _validate_props(combined_props, version)
```

Comment:
@chisholm - per my comment above, could you quickly look over this bit and let me know if I've correctly interpreted how my extra validation should function.

Notably, your change would still raise a `ValueError` if the `_properties` were empty but the `_top_level_properties` were not. I'm not entirely confident that my check against the `combined_properties` is correct or if a spec-complicant construction of a custom extension MUST have a property (ie. the extension type `top-level-property`) if any `_top_level_properties` are defined.

Reply:
`_properties` musn't be empty.  At the very least, you need `extension_type` as a nested property, which describes the type of extension.  Checking for the existence of that particular nested property could be an additional bit of validation I guess.

I suspect we don't really envision people calling `_register_extension()` directly, but if they did, then checking `_toplevel_properties` against `extension_type` might also make sense.  Or maybe toplevel properties are just ignored if extension_type is not "toplevel-property-extension".  Or maybe we produce a warning.  Anyway, if you use the `@CustomExtension` decorator, that should be done correctly for you.

Reply:
Yeah, I didn't mean to suggest that this would be called directly either - it's private so buyer beware IMO.

I think I'd like to keep this change from absorbing too much extra logic for validation if possible and just keep it to what's necessary to enable to checks for reference properties. I think to validate extension properties and potential top-level-properties to make sure they meet section 7.3 or section 11 requirements for custom extensions would need to live in this function anyway so it could be done in a separate MR. Perhaps we can take that on notice if you care to?

What I'm taking from your reply is that I should alter this logic to always ensure that `_properties` is not empty, rather than `combined_props`, which will restore the logic as it currently exists on v3.0.0/master. I'll still call `_validate_props(combined_props)` as the diff currently does, meaning that all property names and `ReferenceProperty` (and variant) type validations will still be performed as usual.

Reply:
> I think I'd like to keep this change from absorbing too much extra logic for validation if possible and just keep it to what's necessary to enable to checks for reference properties. I think to validate extension properties and potential top-level-properties to make sure they meet section 7.3 or section 11 requirements for custom extensions would need to live in this function anyway so it could be done in a separate MR. Perhaps we can take that on notice if you care to?

The extra validation wasn't there before, so evidently no one has so far thought it necessary.  So you wouldn't have to include it with this.

> What I'm taking from your reply is that I should alter this logic to always ensure that `_properties` is not empty, rather than `combined_props`, which will restore the logic as it currently exists on v3.0.0/master. I'll still call `_validate_props(combined_props)` as the diff currently does, meaning that all property names and `ReferenceProperty` (and variant) type validations will still be performed as usual.

Yeah.  I looked to see if the spec had a requirement that a toplevel property extension must define at least one toplevel property, but I couldn't find that.

Reply:
I actually just put that into the new diff since it seemed meaningless to have a top-level-property extension that didn't include anything. Check out the new diff and see what you think.